### PR TITLE
Try in REPL

### LINKED ISSUE #1017: Open in REPL button?
screen shot 2016-12-03 at 11 35 08 pm

---

We have the copy button via https://github.com/babel/babel.github.io/pull/998 now

I was wondering if we could label certain code snippets somehow and then also add a button that opens up a link to the repl with the code that was in the snippet?

Could make http://babeljs.io/docs/learn-es2015/ more interesting

### PR DESCRIPTION
**Pull request by xtuc**
Fixes #1017.

try-magic-icon

Opens a new window with the REPL and the code in it.

### CODE REVIEW

File: _sass/components/_buttons.scss [hunk 1]
```diff
@@ -80,6 +80,18 @@
   }
 }
 
+.btn-open-in-repl {
+  position: absolute;
+  bottom: 0;
+  padding: 1rem;
+  right: 55px; // size of the button.btn-clipboard
```
Comment (xtuc) [inline _sass/components/_buttons.scss]: I'm not a CSS expert. This is causing some issues:
bug.

Would it possible to use a flexbox wrapping both buttons?

File: scripts/open-in-repl.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+/* global jQuery */
+
+(function ($) {
+    'use strict';
+
+    var presets = [ "es2015", "stage-0", "react" ];
```
Comment (xtuc) [inline scripts/open-in-repl.js]: This is meant to be available on the entire website. I tried to be as compatible as possible while not including all presets...

I guess you could use the `package` value in the header to determine which preset is using in the examples.

File: scripts/open-in-repl.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      + '<strong class="btn-icon__label small">Try</strong>'
+      + '</button>'
+    );
+
+    button.on('click', function() {
+      var $code = $(this).parent().find('code');
+
+      var url = '/repl/#'
+              + '?babili=false'
+              + '&evaluate=false'
```
Comment (xtuc) [inline scripts/open-in-repl.js]: Auto-evaluation is disabled since some examples don't produce an output or can't run.

File: scripts/open-in-repl.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      + clipboardIcon
+      + '<strong class="btn-icon__label small">Try</strong>'
+      + '</button>'
+    );
+
+    button.on('click', function() {
+      var $code = $(this).parent().find('code');
+
+      var url = '/repl/#'
+              + '?babili=false'
```
Comment (xtuc) [inline scripts/open-in-repl.js]: We know the source of the documentation (`package_source` in the header). We could enable babili for example from the babili documentation.

File: scripts/open-in-repl.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
+/* global jQuery */
+
+(function ($) {
+    'use strict';
+
+    var presets = [ "es2015", "stage-0", "react" ];
+
+    var selector = '.language-js,.language-javascript';
```
Comment (xtuc) [inline scripts/open-in-repl.js]: This functionality is only for JavaScript code for obvious reasons.

**Comment by satya164** [comment]
May be the triangular run icon is a better icon instead of a magic wand?

**Comment by hzoo** [comment]
Good point about the icon @satya164 - not sure the best icon but yeah maybe a play/run icon is more clear. Or we should just increase the text to say "try it out"

**Comment by existentialism** [comment]
What about moving the actions into the header?

image

It solves the issue of dealing with wrapping around copy/repl links on the last few lines of code.

File: scripts/open-in-repl.js (hunk 4 above)
Comment (existentialism) [inline scripts/open-in-repl.js]: Yeah, decorating a page with `package_source` as a class (i.e. `doc--babel`, `doc--babili`) is probably good first step. We could even enable more granular specific repl settings via convention or a front-matter property.

File: _config.yml [hunk 6]
```diff
@@ -34,6 +34,7 @@ default_js:
   - "https://cdn.jsdelivr.net/docsearch.js/2/docsearch.min.js"
   - "/scripts/search.js"
   - "/scripts/clipboard.js"
+  - "/scripts/open-in-repl.js"
```
Comment (existentialism) [inline _config.yml]: To avoid scanning and modifying the same code blocks twice, we may want to consider combining clipboard/open-in-repl?

File: scripts/open-in-repl.js (hunk 4 above)
Comment (xtuc) [inline scripts/open-in-repl.js]: Yes that would be great. I added the comment as a reminder because it will be a separate PR.

File: _config.yml (hunk 6 above)
Comment (xtuc) [inline _config.yml]: I don't know, I like the separation of responsibility here.

We do update the same JavaScript code blocks twice but I think it's not a big deal.

**Comment by xtuc** [comment]
@existentialism I personally prefer the button in the right corner.

@satya164 @hzoo Maybe we could just write "Try in REPL" without an icon?

**Comment by satya164** [comment]
@xtuc an icon still looks nice IMO

**Comment by xtuc** [comment]
@satya164 Sure. Play button might be common for that.

**Comment by existentialism** [comment]
I'm fine with keeping it in the lower right-hand corner, though I'd definitely increase the space between them.

Also, we'll need to fix cases like:

image

**Comment by xtuc** [comment]
@existentialism Yes, it might be better in the header. I was worried about the size of the buttons in it.

**Comment by xtuc** [comment]
try-in-repl-buttons

@satya164 Yes, the play icon looks better

**Comment by hzoo** [comment]
Nice that looks good!

**Comment by satya164** [comment]
Looks nice 👍

**Comment by hzoo** [comment]
Oh I do like it in the header as well!

**Comment by existentialism** [comment]
@xtuc @hzoo did a quick example at https://github.com/existentialism/babel.github.io/tree/1104-experiment, thoughts?

**Comment by hzoo** [comment]
😍

**Comment by xtuc** [comment]
I also :heart_eyes:. @existentialism could I steal you that?

**Comment by existentialism** [comment]
@xtuc of course, I did a little bit of testing in some browsers, but could def use a second set of eyes.

**Comment by xtuc** [comment]
LGTM. @hzoo ok for merge?

**Comment by hzoo** [comment]
👍 just need to fix copy button

**Comment by hzoo** [comment]
This is awesome! Thanks everyone for the feedback
